onset_maproom/onset_indices.py

```diff
@@ -0,0 +1,115 @@
+import numpy as np
+import xarray as xr
+from numpy.lib.stride_tricks import sliding_window_view
+
+np.random.seed(0)
+
+
+def onset_indices(
+    rain, rainy_day, running_days, running_total, min_rainy_days, dry_days, dry_spell
+):
+    n = rain.shape[0]
+    rain = np.pad(
+        rain,
+        (0, running_days + dry_spell + dry_days - 2),
+        mode="constant",
+        constant_values=0,
+    )
+    rainy = rain > rainy_day
+    first_rainy = np.apply_along_axis(
+        lambda v: np.where(v)[0][0] if len(np.where(v)[0]) != 0 else running_days,
+        1,
+        sliding_window_view(rainy, window_shape=running_days),
+    )
+    wet_spells = (
+        np.convolve(rain, np.ones(running_days, dtype=int), "valid") > running_total
+    ) & (np.convolve(rainy, np.ones(running_days, dtype=int), "valid") > min_rainy_days)
+    dry_spells = np.convolve(~rainy, np.ones(dry_days, dtype=int), "valid") == dry_days
+    dry_spells_ahead = (
+        np.convolve(dry_spells, np.ones(running_days + dry_spell, dtype=int), "valid")
+        != 0
+    )
+    onset_ids = np.where((wet_spells[:n]) & (~dry_spells_ahead))[0]
+    onset_ids = np.unique(first_rainy[onset_ids] + onset_ids)
+    return onset_ids
+
+
+def onset_index(
+    data,
+    data_flag,
+    wlength,
+    wthresh,
+    flagged_thresh,
+    fwlength,
+    false_search,
+    coord_index,
+):
+    """Finds the first window of size wlength where data exceeds wthresh,
+    with at least flagged_thresh count of flagged data (by data_flag),
+    not followed by a false_window of size fwlength of unflagged data,
+    for the following false_search indices
+    returns the index of first flagged data in that window
+    (how much wood would a woodchuck chuck if a woodchuck could chuck wood?)
+    """
+    flagged = data > data_flag
```

Comment:
Another `>` that should be `>=`?

Reply:
not that one because a user might want to be able to define a wet day as anything above 0.

Reply:
Makes sense, and now I see that the Ingrid [docs](https://iridl.ldeo.columbia.edu/dochelp/Documentation/details/index.html?func=onsetDate) say that quite explicitly.

I was misled by a different part of the Ingrid docs:

> running window that totals runningTotals (in prcp units) or more of precipitation

Now I see that this is about the total precip over the window, whereas this line is about the precip for a single day. But isn't it confusing that one precip threshold defines an open interval and the other a closed interval?

Reply:
yes. That is the definition that was given... That is to say the threshold to validate wet to be inclusive...